### PR TITLE
Support moving NotDelivered boxes back InStock

### PR DESCRIPTION
**Pull request by pylipp**
This PR brings the new BoxState `NotDelivered`, and the new mutation `moveNotDeliveredBoxesInStock`.
This mutation can be applied by both sides of a shipment (see trello card).

@vahidbazzaz can you check the updates I made to the FE code especially? I'm not sure if I got everything right in `/views/Box`.

### CODE REVIEW

**Comment by pylipp** [comment]
[2.0 Support moving NotDelivered boxes back InStock](https://trello.com/c/xAkbmzOI/1234-20-support-moving-notdelivered-boxes-back-instock)

File: back/boxtribute_server/business_logic/box_transfer/shipment/mutations.py [hunk 1]
```diff
@@ -72,3 +73,9 @@ def resolve_mark_shipment_as_lost(*_, id):
     )
     authorize(permission="shipment_detail:write")
     return mark_shipment_as_lost(id=id, user=g.user)
+
+
+@mutation.field("moveNotDeliveredBoxesInStock")
+def resolve_move_not_delivered_boxes_in_stock(*_, box_ids):
+    # authorize
```
[the remark this answers is not in the document]
Comment (pylipp) [reply back/boxtribute_server/business_logic/box_transfer/shipment/mutations.py:80]: indeed :innocent:

**Review by vahidbazzaz — APPROVED** [review]
**Functional Review**

1. The box state is correctly changed to NotDelivered when clicking on the trash icon or "No Delivery" button in reconciliation UI.
2. All actions in BoxView are disabled when the box state is NotDelivered
3. When scanning the QR label of the NotDelivered box at the target organization, the error message you don't have permission to view this box appears. If I understand correctly, the correct flow is to display the box view but let the user mark it as found (which is not yet implemented in FE).
4. With QR Scanner and scanning QR label of NotDelivered box on source organization, shows boxview with disabled actions

Thanks @pylipp, as far as I can tell, everything is fine! @aerinsol  just not sure about 3 and whether we should also implement the Marked as found feature in FE

**Comment by pylipp** [comment]
@vahidbazzaz re 3., can you describe exactly what you did on which page; and also paste the errors from the back-end docker-compose service?

**Comment by vahidbazzaz** [comment]
@pylipp I should correct my previous comment. When I used the QR label on the NotDelivered box on the source organization and the QR scanner on the target organization, I got the permission error "You do not have permission to access this box!" in the front end, and the following error in the backend console:

```
2023-10-05 18:29:07 192.168.65.1 - - [05/Oct/2023 14:59:07] "POST /graphql HTTP/1.1" 200 -
2023-10-05 18:29:09 
2023-10-05 18:29:09 
2023-10-05 18:29:09 GraphQL request:18:5
2023-10-05 18:29:09 17 |     code
2023-10-05 18:29:09 18 |     box {
2023-10-05 18:29:09    |     ^
2023-10-05 18:29:09 19 |       ...BoxBasicFields
2023-10-05 18:29:09 Traceback (most recent call last):
2023-10-05 18:29:09   File "/usr/local/lib/python3.10/site-packages/graphql/execution/execute.py", line 521, in execute_field
2023-10-05 18:29:09     result = resolve_fn(source, info, **args)
2023-10-05 18:29:09   File "/app/back/boxtribute_server/business_logic/warehouse/qr_code/fields.py", line 19, in resolve_qr_code_box
2023-10-05 18:29:09     authorize(permission="stock:read", base_id=box.location.base_id)
2023-10-05 18:29:09   File "/app/back/boxtribute_server/authz.py", line 43, in authorize
2023-10-05 18:29:09     return _authorize(*args, **kwargs)
2023-10-05 18:29:09   File "/app/back/boxtribute_server/authz.py", line 117, in _authorize
2023-10-05 18:29:09     raise Forbidden(reason=f"{resource}={value}")
2023-10-05 18:29:09 boxtribute_server.exceptions.Forbidden
2023-10-05 18:29:09 
2023-10-05 18:29:09 The above exception was the direct cause of the following exception:
2023-10-05 18:29:09 
2023-10-05 18:29:09 Traceback (most recent call last):
2023-10-05 18:29:09   File "/usr/local/lib/python3.10/site-packages/graphql/execution/execute.py", line 521, in execute_field
2023-10-05 18:29:09     result = resolve_fn(source, info, **args)
2023-10-05 18:29:09   File "/app/back/boxtribute_server/business_logic/warehouse/qr_code/fields.py", line 19, in resolve_qr_code_box
2023-10-05 18:29:09     authorize(permission="stock:read", base_id=box.location.base_id)
2023-10-05 18:29:09   File "/app/back/boxtribute_server/authz.py", line 43, in authorize
2023-10-05 18:29:09     return _authorize(*args, **kwargs)
2023-10-05 18:29:09   File "/app/back/boxtribute_server/authz.py", line 117, in _authorize
2023-10-05 18:29:09     raise Forbidden(reason=f"{resource}={value}")
2023-10-05 18:29:09 graphql.error.graphql_error.GraphQLError: 
2023-10-05 18:29:09 
2023-10-05 18:29:09 GraphQL request:18:5
2023-10-05 18:29:09 17 |     code
2023-10-05 18:29:09 18 |     box {
2023-10-05 18:29:09    |     ^
2023-10-05 18:29:09 19 |       ...BoxBasicFields
2023-10-05 18:29:09 192.168.65.1 - - [05/Oct/2023 14:59:09] "POST /graphql HTTP/1.1" 200 -
```

Screenshot 2023-10-05 at 6 29 14 PM

Scanning the **NotDelivered** box in the target organization redirects me to the box view.

Screenshot 2023-10-05 at 6 27 27 PM

**Comment by HaGuesto** [comment]
@vahidbazzaz @HaGuesto is this ready to be merged?

**Comment by pylipp** [comment]
@HaGuesto merged now after implementing the missing authz, fixing InTransit/Receiving box accessing through QR code, logging state changes, and fixing conflict with master :sweat_smile:

**Pull request merged**